src/util/init.js

```diff
@@ -38,32 +38,40 @@
      *   http://www.ecse.rpi.edu/Homepages/wrf/Research/Short_Notes/pnpoly.html
      * @param {geo.screenPosition} point The test point
      * @param {geo.screenPosition[]} outer The outer boundary of the polygon
-     * @param {geo.screenPosition[][]?} inner Inner boundaries (holes)
+     * @param {geo.screenPosition[][]} [inner] Inner boundaries (holes)
+     * @param {Object} [range] If specified, range.min.x, range.min.y,
+     *   range.max.x, and range.max.y specified the extents of the outer
+     *   polygon and are used for early detection.
+     * @returns {boolean} true if the point is inside the polygon.
      */
-    pointInPolygon: function (point, outer, inner) {
-      var inside = false, n = outer.length;
+    pointInPolygon: function (point, outer, inner, range) {
+      var inside = false, n = outer.length, i, j;
+
+      if (range && range.min && range.max) {
+        if (point.x < range.min.x || point.y < range.min.y ||
+            point.x > range.max.x || point.y > range.max.y) {
+          return;
+        }
+      }
 
       if (n < 3) {
         // we need 3 coordinates for this to make sense
         return false;
       }
 
-      outer.forEach(function (vert, i) {
-        var j = (n + i - 1) % n;
-        var intersect = (
-          ((outer[i].y > point.y) !== (outer[j].y > point.y)) &&
-          (point.x < (outer[j].x - outer[i].x) *
-                     (point.y - outer[i].y) /
-                     (outer[j].y - outer[i].y) + outer[i].x)
-        );
-        if (intersect) {
+      for (i = 0, j = n - 1; i < n; j = i, i += 1) {
+        if (((outer[i].y > point.y) !== (outer[j].y > point.y)) &&
+            (point.x < (outer[j].x - outer[i].x) *
+            (point.y - outer[i].y) / (outer[j].y - outer[i].y) + outer[i].x)) {
           inside = !inside;
         }
-      });
+      }
 
-      (inner || []).forEach(function (hole) {
-        inside = inside && !geo.util.pointInPolygon(point, hole);
-      });
+      if (inner && inside) {
+        (inner || []).forEach(function (hole) {
```

Comment:
There usually aren't that many holes, so the speed increase in this change
is minor, and I didn't bother to make it.

On Sat, Jun 25, 2016 at 12:53 AM, Aashish Chaudhary <
notifications@github.com> wrote:

> In src/util/init.js
> <https://github.com/OpenGeoscience/geojs/pull/594#discussion_r68484109>:
>
> >
> > -      (inner || []).forEach(function (hole) {
> > -        inside = inside && !geo.util.pointInPolygon(point, hole);
> > -      });
> > +      if (inner && inside) {
> > +        (inner || []).forEach(function (hole) {
>
> since for is much faster than forEach why not use for here as well?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenGeoscience/geojs/pull/594/files/67fa92de4928f8d849ff0abbbb2b8dabbe14eb1e#r68484109>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AIX_RzDWzZTRVoza9g0EWmiY-yWFp9Ugks5qPLRDgaJpZM4I3hEE>
> .
>



-- 
David Manthey
R&D Engineer
Kitware Inc.
(518) 881-4439